### PR TITLE
Refactoring app.spec.tsx from Enzyme to React Testing Library

### PR DESCRIPTION
**Pull request by nathanstilwell**

### CODE REVIEW

**Review by laurenbarker — COMMENTED** [review]
I added a couple nits/questions based on things we've learned in `managed-service`, but this looks great!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-:-N6oQeof9AXoTCnEi41P:b-ktnory)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @laurenbarker and @nathanstilwell)

___
*[`pkg/ui/workspaces/db-console/src/app.spec.tsx` line 53 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6oMRzq6d44ZeHRB7KG:-N6oMRzq6d44ZeHRB7KH:b-rl8nl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6115a48ef1b8952bc221ee176f3e9dc3cada9e8/pkg/ui/workspaces/db-console/src/app.spec.tsx#L53)):*
> ```XML
>     test("routes to <ClusterOverview> component", () => {
>       navigateToPath("/");
>       expect(screen.getByText(/Capacity Usage/)).toBeInTheDocument();
> ```

`toBeInTheDocument()` is a `jest-dom` helper that doesn't provide any value, IMO. It will return the exact same error message as `screen.getByText(/Capacity Usage/);` if the text cannot be found. I think using it could be misleading to developers who are onboarding, since they would assume that the helper was providing some value, and therefore have more difficulty understanding the output of the `react-testing-library` queries.  
___
*[`pkg/ui/workspaces/db-console/src/app.spec.tsx` line 66 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6oPD4C0BLNmzJJtVGV:-N6oPD4C0BLNmzJJtVGW:bm23g27) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6115a48ef1b8952bc221ee176f3e9dc3cada9e8/pkg/ui/workspaces/db-console/src/app.spec.tsx#L66)):*
> ```XML
>     test("routes to <ClusterOverview> component", () => {
>       navigateToPath("/overview");
>       expect(screen.getByText(/Capacity Usage/)).toBeInTheDocument();
> ```

One thing that we do in managed-service that has worked well for us is to define all of the string constants at the top. Assigning a name to the string gives it more context and has the added bonus that when the code changes we only need to update the text value in one location. In some situations we define the constants in a third file and import them into multiple test files to keep everything consistent.

For example,  I don't know what `Capacity usage` represents, but something like, `const clusterCapacityUsageLabel = "Capacity Usage";`, would allow anyone reading the test to understand what the text represents. It's especially helpful for buttons with generic text :smile: 
___
*[`pkg/ui/workspaces/db-console/src/app.spec.tsx` line 79 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6oOtx-76wCFYFBddjJ:-N6oOtx-76wCFYFBddjK:b-y6kgqm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6115a48ef1b8952bc221ee176f3e9dc3cada9e8/pkg/ui/workspaces/db-console/src/app.spec.tsx#L79)):*
> ```XML
>     test("routes to <NodeList> component", () => {
>       navigateToPath("/overview");
>       expect(screen.getByText(/Capacity Usage/)).toBeInTheDocument();
> ```

Can this be a regular string instead of a regex? Similar question for all of the text queries that aren't using any regex functionality.
___
*[`pkg/ui/workspaces/db-console/src/app.spec.tsx` line 282 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6oO9gw-luBTEyW-pcp:-N6oO9gx19F-a798Twx8:brpgz6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6115a48ef1b8952bc221ee176f3e9dc3cada9e8/pkg/ui/workspaces/db-console/src/app.spec.tsx#L282)):*
> ```XML
>     test("routes to <DatabaseDetailsPage> component", () => {
>       navigateToPath("/database/some-db-name");
>       const result = document.querySelector(
> ```

Is this query selector necessary? (ditto for all the other uses of `querySelector`)

File: pkg/ui/workspaces/db-console/src/views/app/components/errorMessage/errorBoundary.tsx [hunk 1]
```diff
@@ -41,7 +41,9 @@ export default class ErrorBoundary extends React.Component<
 
   componentDidCatch(error: Error, errorInfo: ErrorInfo) {
     // Console.error for developer visibility.
-    console.error(error);
+    console.error("[ErrorBoundary::componentDidCatch] error = ", error);
+    console.error("[ErrorBoundary::componentDidCatch] errorInfo = ", errorInfo);
+    console.log("children = ", this.props.children);
```
Comment (sjbarag) [inline pkg/ui/workspaces/db-console/src/views/app/components/errorMessage/errorBoundary.tsx:46]: We should revert this, it was just debugging code and will cause a lot of spurious logging for end-users :)

File: pkg/ui/workspaces/db-console/src/redux/history.ts [hunk 2]
```diff
@@ -7,6 +7,7 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0, included in the file
 // licenses/APL.txt.
+/* globals module */
```
Comment (sjbarag) [inline pkg/ui/workspaces/db-console/src/redux/history.ts:10]: I don't remember if this was me or you but 🤷 I think I goofed.

File: pkg/ui/workspaces/db-console/src/app.spec.tsx [hunk 3]
```diff
@@ -25,678 +25,706 @@ stubComponentInModule("src/views/statements/statementsPage", "default");
...
   });
 
   const navigateToPath = (path: string) => {
     history.push(path);
-    appWrapper.update();
   };
 
   describe("'/' path", () => {
-    it("routes to <ClusterOverview> component", () => {
+    test("routes to <ClusterOverview> component", () => {
```
Comment (sjbarag) [inline pkg/ui/workspaces/db-console/src/app.spec.tsx:51]: Any reason to change these from `it` to `test`?  One is aliased to the other in Jest (though I can't remember which one), so they should be equivalent.

**Review by sjbarag — COMMENTED** [review]
I found some things I forgot to undo whooooooops

**Review by nathanstilwell — COMMENTED** [review]
Thank you both for the review!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-:-N6y2bRo6fizO0ZBp3Bs:b-unmja)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @laurenbarker and @sjbarag)

___
*[`pkg/ui/workspaces/db-console/src/redux/history.ts` line 10 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6jrxuhcF3vjwE0cdP0-r5-10:-N6y2ovz54pusWzgtqol:b-zbhzkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe64a31d9b9f56152b01af8d6e554586b2f28b78/pkg/ui/workspaces/db-console/src/redux/history.ts#L10)):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

I don't remember if this was me or you but 🤷 I think I goofed.
</blockquote></details>

That looks like me, and is definitely a useless left-over. Good catch.
___
*[`pkg/ui/workspaces/db-console/src/views/app/components/errorMessage/errorBoundary.tsx` line 46 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6jrxuhcF3vjwE0cdP2-r5-46:-N6y2yQcDsPOXcp-BPC1:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe64a31d9b9f56152b01af8d6e554586b2f28b78/pkg/ui/workspaces/db-console/src/views/app/components/errorMessage/errorBoundary.tsx#L46)):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

We should revert this, it was just debugging code and will cause a lot of spurious logging for end-users :)
</blockquote></details>

:+1: 
___
*[`pkg/ui/workspaces/db-console/src/app.spec.tsx` line 51 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6jrxui2AAM-RYEE-sf-r5-51:-N6y354ZETIFz2fj_97L:b-ce95uf) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe64a31d9b9f56152b01af8d6e554586b2f28b78/pkg/ui/workspaces/db-console/src/app.spec.tsx#L51)):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

Any reason to change these from `it` to `test`?  One is aliased to the other in Jest (though I can't remember which one), so they should be equivalent.
</blockquote></details>

`it` is aliased to `test`. `test` is used in the Jest documentation and I haven't heard very much discussion about the BDD styles of testing, but in my mind the point of using `it` is to write tests in the form of `it ("should do a thing", ...);` Since these (and many other tests in the codebase) don't use this form, I thought the more generic `test` was a better choice for us so I was trying to model it here.

Please let me know if you feel differently. I wouldn't mind a conversation. 
___
*[`pkg/ui/workspaces/db-console/src/app.spec.tsx` line 53 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6oMRzq6d44ZeHRB7KG:-N6y4yWY4dSC1bmw7JmK:bcdo1jj) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe64a31d9b9f56152b01af8d6e554586b2f28b78/pkg/ui/workspaces/db-console/src/app.spec.tsx#L53)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

`toBeInTheDocument()` is a `jest-dom` helper that doesn't provide any value, IMO. It will return the exact same error message as `screen.getByText(/Capacity Usage/);` if the text cannot be found. I think using it could be misleading to developers who are onboarding, since they would assume that the helper was providing some value, and therefore have more difficulty understanding the output of the `react-testing-library` queries.  
</blockquote></details>

Thanks for the feedback Lauren. I came across `jest-dom` in the Testing Library docs and added it here. I take your point though, so would we (in this case) remove the `expect` statement entirely? Like,

```
navigateToPath("/");
screen.getByText(/Capacity Usage/);
```

?
___
*[`pkg/ui/workspaces/db-console/src/app.spec.tsx` line 66 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6oPD4C0BLNmzJJtVGV:-N6y5irrEKrLCO_UVuUM:bpxaonh) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe64a31d9b9f56152b01af8d6e554586b2f28b78/pkg/ui/workspaces/db-console/src/app.spec.tsx#L66)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

One thing that we do in managed-service that has worked well for us is to define all of the string constants at the top. Assigning a name to the string gives it more context and has the added bonus that when the code changes we only need to update the text value in one location. In some situations we define the constants in a third file and import them into multiple test files to keep everything consistent.

For example,  I don't know what `Capacity usage` represents, but something like, `const clusterCapacityUsageLabel = "Capacity Usage";`, would allow anyone reading the test to understand what the text represents. It's especially helpful for buttons with generic text :smile: 
</blockquote></details>

This is a good call out. At some point I remember seeing that in a managed-server test. I'll refactor these test to use constants.
___
*[`pkg/ui/workspaces/db-console/src/app.spec.tsx` line 79 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6oOtx-76wCFYFBddjJ:-N6y7p79B5o78aIdLGSi:b-ykm2o8) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe64a31d9b9f56152b01af8d6e554586b2f28b78/pkg/ui/workspaces/db-console/src/app.spec.tsx#L79)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

Can this be a regular string instead of a regex? Similar question for all of the text queries that aren't using any regex functionality.
</blockquote></details>

Sure. Again, I saw an example in the docs and was using Regex for several of the tests so I just left them in for consistency. Do you think it makes it harder to read?
___
*[`pkg/ui/workspaces/db-console/src/app.spec.tsx` line 282 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6oO9gw-luBTEyW-pcp:-N6y80ywBv_lEzPudDtJ:by9c3w9) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe64a31d9b9f56152b01af8d6e554586b2f28b78/pkg/ui/workspaces/db-console/src/app.spec.tsx#L282)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

Is this query selector necessary? (ditto for all the other uses of `querySelector`)
</blockquote></details>

These components are placeholders loaded by `stubComponentInModule` above as empty Divs with props. I thought this would be roughly the same as using a `data-testid`, but since RTL has support for that do you think it would be better to use test ids in these cases?

**Comment by nathanstilwell** [comment]
*[`pkg/ui/workspaces/db-console/src/app.spec.tsx` line 53 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6oMRzq6d44ZeHRB7KG:-N71ZhfU1LEZeNthgSkz:b-lfju2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe64a31d9b9f56152b01af8d6e554586b2f28b78/pkg/ui/workspaces/db-console/src/app.spec.tsx#L53)):*
<details><summary><i>Previously, nathanstilwell (Nathan Stilwell) wrote…</i></summary><blockquote>

Thanks for the feedback Lauren. I came across `jest-dom` in the Testing Library docs and added it here. I take your point though, so would we (in this case) remove the `expect` statement entirely? Like,

```
navigateToPath("/");
screen.getByText(/Capacity Usage/);
```

?
</blockquote></details>

Also @laurenbarker, do you think there's any value in having `jest-dom` added as a dependency? I believe I only added it for the `toBeInTheDocument` matcher, but if there are no other matchers that are worthwhile, I can remove it.

**Comment by nathanstilwell** [comment]
*[`pkg/ui/workspaces/db-console/src/app.spec.tsx` line 79 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6oOtx-76wCFYFBddjJ:-N71_BF1ARQj0XVBSREY:b-3rx2t1) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe64a31d9b9f56152b01af8d6e554586b2f28b78/pkg/ui/workspaces/db-console/src/app.spec.tsx#L79)):*
<details><summary><i>Previously, nathanstilwell (Nathan Stilwell) wrote…</i></summary><blockquote>

Sure. Again, I saw an example in the docs and was using Regex for several of the tests so I just left them in for consistency. Do you think it makes it harder to read?
</blockquote></details>

I guess after refactoring these to use constants it's kind of a moot point (about readability).

**Review by laurenbarker — APPROVED** [review]
:lgtm:, but it wouldn't hurt to get a second approval from those who actually work in this codebase. @sjbarag probably counts as a `db-console` contributor at this point :) 

> Reviewed 2 of 4 files at r5, 2 of 2 files at r6, 1 of 1 files at r7, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-:-N7Gj4yM8_5nkfPUTyAF:bc5zv2c)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @laurenbarker, @nathanstilwell, and @sjbarag)

___
*[`-- commits` line 165 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N7GifCf2awcVcEDMufr):*
Is this PR only this commit & following commits?
___
*[`pkg/ui/workspaces/db-console/src/app.spec.tsx` line 53 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6oMRzq6d44ZeHRB7KG:-N7Gf_Sj8k2WG3g_liWy:bkkxssh) ([raw file](https://github.com/cockroachdb/cockroach/blob/32c6cdb3995e77d263363635e4d5cc90533370c9/pkg/ui/workspaces/db-console/src/app.spec.tsx#L53)):*
<details><summary><i>Previously, nathanstilwell (Nathan Stilwell) wrote…</i></summary><blockquote>

Also @laurenbarker, do you think there's any value in having `jest-dom` added as a dependency? I believe I only added it for the `toBeInTheDocument` matcher, but if there are no other matchers that are worthwhile, I can remove it.
</blockquote></details>

That's correct! The `expect()` statement can be removed entirely.

`jest-dom` does provide some useful helpers like `toBeEnabled`, `toBeDisabled`, and `toBeInvalid` that we use regularly in `managed-service`. However, most of those are related to forms, so I'm not sure if the same helpers would be used in this repo.
___
*[`pkg/ui/workspaces/db-console/src/app.spec.tsx` line 79 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6oOtx-76wCFYFBddjJ:-N7Gga4e5I29JW8u9Ltb:b-4or9yu) ([raw file](https://github.com/cockroachdb/cockroach/blob/32c6cdb3995e77d263363635e4d5cc90533370c9/pkg/ui/workspaces/db-console/src/app.spec.tsx#L79)):*
<details><summary><i>Previously, nathanstilwell (Nathan Stilwell) wrote…</i></summary><blockquote>

I guess after refactoring these to use constants it's kind of a moot point (about readability).
</blockquote></details>

I don't think it makes it more difficult to read, but it can be confusing when developers who aren't familiar with regexes modify or try to copy the tests. If only the queries that use regex functionality use regex syntax, it's more obvious that regexes have a specific use case.
___
*[`pkg/ui/workspaces/db-console/src/app.spec.tsx` line 282 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6oO9gw-luBTEyW-pcp:-N7GhmeU-SdiJHoobxz-:bcrizcs) ([raw file](https://github.com/cockroachdb/cockroach/blob/32c6cdb3995e77d263363635e4d5cc90533370c9/pkg/ui/workspaces/db-console/src/app.spec.tsx#L282)):*
<details><summary><i>Previously, nathanstilwell (Nathan Stilwell) wrote…</i></summary><blockquote>

These components are placeholders loaded by `stubComponentInModule` above as empty Divs with props. I thought this would be roughly the same as using a `data-testid`, but since RTL has support for that do you think it would be better to use test ids in these cases?
</blockquote></details>

Ohh, the component is mocked for tests? Yeah, `data-testid` would make sense since it's only for tests anyway.

**Comment by nathanstilwell** [comment]
@sjbarag is working his way up to _major contributor_ soon :wink:

**Review by nathanstilwell — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83653)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @laurenbarker, @nathanstilwell, and @sjbarag)

___
*[`-- commits` line 165 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N7GifCf2awcVcEDMufr):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

Is this PR only this commit & following commits?
</blockquote></details>

Yes, since this PR was the last one on top of draft PRs I didn't want to add much more to all of the rebase juggling I was already doing.
___
*[`pkg/ui/workspaces/db-console/src/app.spec.tsx` line 282 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6oO9gw-luBTEyW-pcp:-N7GoW85DUCg0pZXc9Lc:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/32c6cdb3995e77d263363635e4d5cc90533370c9/pkg/ui/workspaces/db-console/src/app.spec.tsx#L282)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

Ohh, the component is mocked for tests? Yeah, `data-testid` would make sense since it's only for tests anyway.
</blockquote></details>

:+1:

**Review by maryliag — APPROVED** [review]
just one small nit form me, otherwise :lgtm: 

> Reviewed 1 of 2 files at r2, 1 of 2 files at r3, 3 of 4 files at r5, 2 of 2 files at r6, 1 of 1 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-:-N7HXbcI8uaPzFWR0GZS:biaubhl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @laurenbarker, @maryliag, @nathanstilwell, and @sjbarag)

___
*[`pkg/ui/workspaces/db-console/src/views/app/components/errorMessage/errorBoundary.tsx` line 46 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6jrxuhcF3vjwE0cdP2-r5-46:-N7HWqJa496FazNgqrpK:b-9m7uwr) ([raw file](https://github.com/cockroachdb/cockroach/blob/32c6cdb3995e77d263363635e4d5cc90533370c9/pkg/ui/workspaces/db-console/src/views/app/components/errorMessage/errorBoundary.tsx#L46)):*
<details><summary><i>Previously, nathanstilwell (Nathan Stilwell) wrote…</i></summary><blockquote>

:+1: 
</blockquote></details>

I was the one who added this. We use it to see errors on Datadog that people are hitting, so this should be kept.

**Review by nathanstilwell — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83653)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @laurenbarker, @maryliag, @nathanstilwell, and @sjbarag)

___
*[`pkg/ui/workspaces/db-console/src/app.spec.tsx` line 53 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83653#-N6oMRzq6d44ZeHRB7KG:-N7GnVF1EHoZzM98AvSx:bmitp08) ([raw file](https://github.com/cockroachdb/cockroach/blob/32c6cdb3995e77d263363635e4d5cc90533370c9/pkg/ui/workspaces/db-console/src/app.spec.tsx#L53)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

That's correct! The `expect()` statement can be removed entirely.

`jest-dom` does provide some useful helpers like `toBeEnabled`, `toBeDisabled`, and `toBeInvalid` that we use regularly in `managed-service`. However, most of those are related to forms, so I'm not sure if the same helpers would be used in this repo.
</blockquote></details>

Since DB Console has very few forms and is considered mostly "read-only" I ended up removing the dependency entirely.

**Pull request merged**